File: dotnet/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/DeliveredState.cs
namespace StateDesignPattern.Example1;

/// <summary>
/// Concrete State - represents an order that has been delivered.
/// </summary>
public class DeliveredState : IOrderState
{
    public void ProcessOrder(Order order)
    {
        Console.WriteLine("Cannot process. Order has been delivered.");
    }

    public void ShipOrder(Order order)
    {
        Console.WriteLine("Cannot ship. Order has been delivered.");
    }

    public void DeliverOrder(Order order)
    {
        Console.WriteLine("Order has already been delivered.");
    }

    public void CancelOrder(Order order)
    {
        Console.WriteLine("Cannot cancel. Order has been delivered.");
    }

    public string GetStateName() => "Delivered";
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/src/DeliveredState.ts
import { IOrderState } from './IOrderState';
import { Order } from './Order';

export class DeliveredState implements IOrderState {
  processOrder(order: Order): void {
    console.log('Cannot process. Order has been delivered.');
  }

  shipOrder(order: Order): void {
    console.log('Cannot ship. Order has been delivered.');
  }

  deliverOrder(order: Order): void {
    console.log('Order has already been delivered.');
  }

  cancelOrder(order: Order): void {
    console.log('Cannot cancel. Order has been delivered.');
  }

  getStateName(): string {
    return 'Delivered';
  }
}
